### PR TITLE
Check if user supplied old style gradients and Hessians and correct  it.

### PR DESCRIPTION
**Pull request by pkofod**

### CODE REVIEW

**Comment by pkofod** [comment]
@tkelman don't know if you have time, but it makes more sense to take the discussion up front, rather than when I'm tagging.

In the next version of Optim (and NLsolve) we're going to change the order of storage and `x` input, such that the previous
```
function gradient!(x, storage)
    calculate g and store it in storage
end
```
is now
```
function gradient!(storage, x)
    ...
end
```
This change checks to see if `x` is mutated instead of `storage`, and switches the input order using an anonymous function if that is the case. This way we can "deprecate" a change that is going to break a lot of code. Any comments?

**Comment by tkelman** [comment]
What do you want me to comment on?

**Comment by pkofod** [comment]
> What do you want me to comment on?

I'm sorry, it's slipped my mind what I wanted you to comment on.

File: src/objective_types.jl [hunk 1]
```diff
@@ -54,14 +86,41 @@ type TwiceDifferentiable{T<:Real} <: AbstractObjective
...
+    _new_fg! = fix_order(g, x_seed, fg!, "fg!")
+
+    local _new_h!
+    try
+        _H = copy(H)
+        _x = copy(x_seed)
+        h!(_H, _x)
+        _new_h! = (storage, x) -> h!(storage, x)
+    catch m
+        if isa(m, MethodError) || isa(m, BoundsError)
```
Comment (tkelman) [inline src/objective_types.jl]: should this proceed silently if anything else goes wrong?

Comment (pkofod) [inline src/objective_types.jl]: no, there should have been a rethrow there, I'll patch it. Thanks.

**Pull request merged**